util/stopper.go

```diff
@@ -88,40 +92,83 @@ func (s *Stopper) AddCloser(c Closer) {
 	s.closers = append(s.closers, c)
 }
 
-// StartTask adds one to the count of tasks left to drain in the
-// system. Any worker which is a "first mover" when starting tasks
-// must call this method before starting work on a new task and must
-// subsequently invoke FinishTask() when the task is complete.
-// First movers include goroutines launched to do periodic work and
-// the kv/db.go gateway which accepts external client
-// requests.
+// A PendingTask is returned by StartTask. For simplicity this is just a
+// function pointer which is nil if the task was refused and is executed
+// when the task is done, but the Ok() and Done() methods defined on it
+// are more idiomatic.
+type PendingTask func() bool
+
+// Ok must be called to confirm that actual work may be started for this task.
+func (pt PendingTask) Ok() bool {
+	return pt != nil
+}
+
+// Done is called (exactly once) when the started work has completed.
+// It does not need to be called when Ok() returned false.
+func (pt PendingTask) Done() {
+	if pt.Ok() {
+		pt()
+	}
+}
+
+// StartTask adds one to the count of tasks left to drain in the system. Any
+// worker which is a "first mover" when starting tasks must call this method
+// before starting work on a new task and must subsequently invoke Done() on
+// the returned handle  when the task is complete. First movers include
+// goroutines launched to do periodic work and the kv/db.go gateway which
+// accepts external client requests.
 //
-// Returns true if the task can be launched or false to indicate the
-// system is currently draining and the task should be refused.
-func (s *Stopper) StartTask() bool {
+// Ok() must be called to confirm that the task can be launched. Ok() returns
+// false to indicate that the system is currently draining and the task should
+// must not proceed. It's not necessary to call Done() in that case.
+func (s *Stopper) StartTask() PendingTask {
```

Comment:
while the functionality is fine here. I feel that this abstraction matches more one of a work permit.

I prefer

permit := stopper.GetPermit() WorkPermit {

if (permit.Issued()) {
  defer permit.Return()
  // do work
}
If you like that, I'd use gorename

token might be another option, but task doesn't seem right, because it's not a task.